text/0249-v2-experiments.md

```diff
@@ -0,0 +1,137 @@
+
+The existing contributor guide does not mention experimental code, so the following will be added after the intro
+paragraph:
+
+> If you are contributing to a 2.x release, keep in mind that you may not introduce breaking changes to any public
+> facing API. Experimental modules are developed in a separate package, and experimental APIs in stable modules are
+> hidden behind a feature flag. [link]
+
+# Design Summary
+
+- Before releasing v2 to GA, all experimental code will either be marked as stable or deprecated and removed from the
+  packages that we vend, such as aws-cdk and aws-cdk-lib.
+- A new folder structure will be created under packages called @aws-cdk-previews. All new modules will begin their
+  lifecycle there, and any modules we decide to deprecate for further experimentation will be moved there.
+- The new @aws-cdk-previews package will take a peer dependency on @aws-cdk.
+- When a module is ready for graduation to GA, in a single commit, the code will be moved to packages/@aws-cdk and the
+  metadata tags for stability and maturity will be set to 'stable'.
+- A feature flag strategy will be devised to handle experimental APIs within stable modules, or in the core. Runtime
+  checks will make sure users who have not opted in do not call experimental APIs.
+
+# Drawbacks
+
+- When we graduate a module, users who have participated in the Developer Preview will have to update their imports to
+  make use of the stable version in the main repository.
+- For developers who are working on experimental code that has complex dependencies with the core and with other
+  modules, development will be more complicated.
+- In some cases, experimentation will not be possible. New features introduced to the main repository will have to be
+  vetted much more carefully before they are merged, since any public API released as part of v2 will have to be
+  supported for the lifetime of the major version.
+- The stable package that we vend cannot take dependencies on the experimental package.
+- API design for L2 construct libraries will have to take into account the possibility of future additions that are
+  implemented in a separate package. This might lead to awkward choices, like making functions public that otherwise
+  would have been private.
+
+# Alternatives
+
+Some alternatives to this strategy are:
+
+- Make no changes and continue to experiment in the main release.
+  - Improve documentation so that it is more obvious to users that they are importing experimental code.
+  - Make it a policy to actually include the name 'experimental' in all experimental APIs.
+- Use clever build tricks to hide experimental modules. This might be possible in Typescript, where the type files are
+  separate from implementation, but might prove difficult or impossible in other languages such as Python.
+- Call the experimental package @aws-cdk-experiments instead of @aws-cdk-previews.
+
```

Comment:
Where is the alternative to separate experimental code into *submodules* within `aws-cdk-lib`.

Something like this:

```ts
import * as eks from 'aws-cdk-lib/aws-eks-experimental';
// or we could do
import * as eks from 'aws-cdk-lib/experimental/aws-eks';
```

The benefit of this approach is that we will still vend everything as a single module so we can avoid the potential risk of peer-deps hell

Reply:
Not sure but didn't we have a strong reaction from others against submodules?

Reply:
#What kind of reaction? (I mean submodules within `AWS-CDK-lib`, not git submodules)

Reply:
Adding this alternative to the RFC. If we can be sure we won't end up shooting ourselves in the foot, we should definitely explore the idea.